vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py

```diff
@@ -3,20 +3,36 @@
 from typing import Optional
 
 import torch
+from flashinfer.comm.trtllm_alltoall import MnnvlMoe, MoEAlltoallInfo
 
+import vllm.envs as envs
 import vllm.model_executor.layers.fused_moe.modular_kernel as mk
-from vllm.distributed import get_dp_group
+from vllm.distributed import get_dp_group, get_ep_group
+from vllm.distributed.device_communicators.base_device_communicator import (
+    All2AllManagerBase)
 from vllm.forward_context import get_forward_context
 from vllm.model_executor.layers.fused_moe.config import FusedMoEQuantConfig
 from vllm.model_executor.layers.fused_moe.utils import (
     moe_kernel_quantize_input)
 from vllm.utils.flashinfer import nvfp4_block_scale_interleave
 
 
+def get_global_num_tokens_cpu():
+    cu_sizes = get_forward_context().dp_metadata.cu_tokens_across_dp_cpu
+    sizes = [cu_sizes[0].item()]
+    for i in range(1, len(cu_sizes)):
+        sizes.append((cu_sizes[i] - cu_sizes[i - 1]).item())
+    return sizes
+
+
 def get_local_sizes():
     return get_forward_context().dp_metadata.get_chunk_sizes_across_dp_rank()
 
 
+enable_flashinfer_alltoall = envs.VLLM_ALL2ALL_BACKEND == "flashinfer"
+enable_flashinfer_fp4_allgather = not enable_flashinfer_alltoall
```

Comment:
Why not have two separate `PrepareaAndFinalize` classes if these conditions are mutually exclusive?